### PR TITLE
Adds Hair Dye

### LINKED ISSUE #12074: Hairdye
So nerds can change their hair color in round without admin intervention and without genetics

### PR DESCRIPTION
**Pull request by Shadowmech88**
Closes #12074

:cl:
 * rscadd: Adds a can of sprayable hair dye, which can make hair into any color. Target the mouth to color facial hair.
 * rscadd: The standard costume crate now comes with a can of hair dye.

### CODE REVIEW

File: code/game/objects/items/weapons/cosmetics.dm [hunk 1]
```diff
@@ -88,4 +88,87 @@
...
+		color_b = hex2num(copytext(new_color, 6, 8))
+	update_icon()
+
+/obj/item/weapon/hair_dye/attack(mob/M as mob, mob/user as mob)
+	if(!istype(M, /mob))
+		return
+
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		if(isjusthuman(H) || isunathi(H))	//Humans and unathi are the only races with actual hair.
```
Comment (ComicIronic) [inline code/game/objects/items/weapons/cosmetics.dm]: Not OO, species should be declaring whether or not they have hair.

File: code/game/objects/items/weapons/cosmetics.dm [hunk 2]
```diff
@@ -88,4 +88,87 @@
...
+			if(area == "mouth")
+				if(H.f_style == "Shaved")	//if they have no facial hair
+					to_chat(user, "<span class='notice'>[H == user ? "You don't" : "\The [H] doesn't"] seem to have any facial hair!</span>")
+					return
+			else
+				if(H.h_style == "Bald")	//if they have no hair
+					to_chat(user, "<span class='notice'>[H == user ? "You don't" : "\The [H] doesn't"] seem to have any hair!</span>")
+					return
+			if(H == user)
+				user.visible_message("<span class='notice'>[user] colors their [area == "mouth" ? "facial hair" : "hair"] with \the [src].</span>", \
```
Comment (ComicIronic) [inline code/game/objects/items/weapons/cosmetics.dm]: Just take the result of this hair/facial hair conditional once and use it everywhere.

**Review by ComicIronic — DISMISSED** [review]

File: code/game/objects/items/weapons/cosmetics.dm (hunk 1 above)
Comment (Shadowmech88) [reply code/game/objects/items/weapons/cosmetics.dm]: To my knowledge, there doesn't exist any flag or variable that governs whether a species can have hair.

Comment (clusterfack) [inline code/game/objects/items/weapons/cosmetics.dm]: Are you sure? I'm pretty sure there is, at the very least its used in icon updating

Comment (Shadowmech88) [reply code/game/objects/items/weapons/cosmetics.dm]: I looked at the update_icons code.

Apparently each individual hairstyle has a list of species that are allowed to have it. It checks what hairstyle the mob has, then checks if the mob's species is in that hairstyle's list of allowed species.

Comment (ComicIronic) [reply code/game/objects/items/weapons/cosmetics.dm]: Then create that variable.

Comment (Shadowmech88) [reply code/game/objects/items/weapons/cosmetics.dm]: Done.

Comment (Intigracy) [reply code/game/objects/items/weapons/cosmetics.dm]: https://github.com/d3athrow/vgstation13/blob/58346d73cc2db766b895efc12e66ab7619d21917/code/modules/mob/living/carbon/human/update_icons.dm#L314

Comment (Intigracy) [reply code/game/objects/items/weapons/cosmetics.dm]: https://github.com/d3athrow/vgstation13/blob/58346d73cc2db766b895efc12e66ab7619d21917/code/modules/mob/living/carbon/human/update_icons.dm#L324

**Comment by Shadowmech88** [comment]
@ComicIronic 
I changed the species check to a better one that checks whether the mob's particular hairstyle allows coloring via the do_colouration var.

**Comment by SonixApache** [comment]
Standard Costume Crates are locked to theater access, by the way.

**Comment by Shadowmech88** [comment]
Why does this have the "compile failure" tag? It's showing "All checks have passed" for me.

File: code/game/objects/items/weapons/cosmetics.dm [hunk 3]
```diff
@@ -88,4 +88,97 @@
...
+
+/obj/item/weapon/hair_dye/attack(mob/M as mob, mob/user as mob)
+	if(!istype(M, /mob))
+		return
+
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		var/area = user.zone_sel.selecting
+		var/area_string = "hair"
+		if(area == "mouth")
```
Comment (Intigracy) [inline code/game/objects/items/weapons/cosmetics.dm:133]: Only thing I can think of at this point is checking if they have their hair/face covered.

**Review by Intigracy — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/cosmetics.dm (hunk 3 above)
Comment (Shadowmech88) [reply code/game/objects/items/weapons/cosmetics.dm:133]: Done.

File: code/game/objects/items/weapons/cosmetics.dm [hunk 4]
```diff
@@ -131,23 +131,31 @@
...
+				to_chat(user, "<span class='notice'>You can't color [H == user ? "your" : "\the [H]'s"] hair through that [cover.name]!</span>")
+				return
 			if(!H.h_style || H.h_style == "Bald")	//if they have no hair
 				to_chat(user, "<span class='notice'>[H == user ? "You don't" : "\The [H] doesn't"] seem to have any hair!</span>")
 				return
 			else
 				var/datum/sprite_accessory/hair_style = hair_styles_list[H.h_style]
 				if(!hair_style.do_colouration)
-					to_chat(user, "<span class='notice'>\The [H] doesn't seem to have any colorable hair!</span>")
+					to_chat(user, "<span class='notice'>\The [H == user ? "You don't" : "\The [H] doesn't"] doesn't seem to have any colorable hair!</span>")
```
Comment (Intigracy) [inline code/game/objects/items/weapons/cosmetics.dm]: Both of these say \The You or \The \The [H].

**Review by Intigracy — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/cosmetics.dm (hunk 4 above)
Comment (Intigracy) [reply code/game/objects/items/weapons/cosmetics.dm]: By both I mean this and 144

**Comment by Shadowmech88** [comment]
Typos fixed.

**Review by Intigracy — APPROVED** [review]

**Comment by Shadowmech88** [comment]
Do I need @ComicIronic to personally approve his review? Because I did make the changes he wanted.

File: code/game/objects/items/weapons/cosmetics.dm [hunk 5]
```diff
@@ -88,4 +88,105 @@
...
+
+/obj/item/weapon/hair_dye/attack(mob/M as mob, mob/user as mob)
+	if(!istype(M, /mob))
+		return
+
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		var/area = user.zone_sel.selecting
+		var/area_string = "hair"
+		if(area == "mouth")
```
Comment (PJB3005) [inline code/game/objects/items/weapons/cosmetics.dm:133]: This should all be defines.

**Comment by ihadtoregisterforthis** [comment]
Is it instantaneous or does it take time to apply?

**Comment by Shadowmech88** [comment]
It's instant if you're applying it to yourself, otherwise it has the same delay as applying lipstick to someone else.

**Pull request merged**